tests/test_separablePipelineExecutor.py

```diff
@@ -0,0 +1,548 @@
+        self.assertTrue(graph.isConnected)
+        self.assertEqual(len(graph), 1)
+        self.assertEqual({q.taskDef.label for q in graph.inputQuanta}, {"b"})
+        self.assertEqual({q.taskDef.label for q in graph.outputQuanta}, {"b"})
+
+    def test_make_graph_nowhere_skippartial_clobber(self):
+        executor = SeparablePipelineExecutor(self.butler, skip_existing=True, clobber_output=True)
+        pipeline = Pipeline.fromFile(self.pipeline_file)
+
+        self.butler.put({"zero": 0}, "input")
+        self.butler.put({"zero": 0}, "intermediate")
+
+        graph = executor.make_graph(pipeline)
+        self.assertFalse(graph.isConnected)  # Both tasks run, but can use old values for b
+        self.assertEqual(len(graph), 2)
+        self.assertEqual({q.taskDef.label for q in graph.inputQuanta}, {"a", "b"})
+        self.assertEqual({q.taskDef.label for q in graph.outputQuanta}, {"a", "b"})
+
+    def test_make_graph_noinput(self):
+        executor = SeparablePipelineExecutor(self.butler)
+        pipeline = Pipeline.fromFile(self.pipeline_file)
+
+        graph = executor.make_graph(pipeline)
+        self.assertEqual(len(graph), 0)
+
+    def test_make_graph_alloutput_skip(self):
+        executor = SeparablePipelineExecutor(self.butler, skip_existing=True)
+        pipeline = Pipeline.fromFile(self.pipeline_file)
+
+        butlerTests.addDatasetType(self.butler, "output", set(), "StructuredDataDict")
+        butlerTests.addDatasetType(self.butler, "b_log", set(), "ButlerLogRecords")
+        butlerTests.addDatasetType(self.butler, "b_metadata", set(), "TaskMetadata")
+
+        self.butler.put({"zero": 0}, "input")
+        self.butler.put({"zero": 0}, "intermediate")
+        self.butler.put(lsst.daf.butler.ButlerLogRecords.from_records([]), "a_log")
+        self.butler.put(TaskMetadata(), "a_metadata")
+        self.butler.put({"zero": 0}, "output")
+        self.butler.put(lsst.daf.butler.ButlerLogRecords.from_records([]), "b_log")
+        self.butler.put(TaskMetadata(), "b_metadata")
+
+        graph = executor.make_graph(pipeline)
+        self.assertEqual(len(graph), 0)
+
+    def test_run_pipeline_noskip_noclobber_fullgraph(self):
+        executor = SeparablePipelineExecutor(self.butler, skip_existing=False, clobber_output=False)
+        pipeline = Pipeline.fromFile(self.pipeline_file)
+        self.butler.put({"zero": 0}, "input")
+        graph = executor.make_graph(pipeline)
+        executor.pre_execute_qgraph(
+            graph,
+            register_dataset_types=True,
+            save_init_outputs=True,
+            save_versions=False,
+        )
+
+        executor.run_pipeline(graph)
+        self.butler.registry.refresh()
+        self.assertEqual(self.butler.get("intermediate"), {"zero": 0, "one": 1})
+        self.assertEqual(self.butler.get("output"), {"zero": 0, "one": 1, "two": 2})
+
+    # TODO: run_pipeline should do nothing on an empty graph, but how do you
+    # prove a negative?
```

Comment:
Could check that the data repository is in the same state before and after, e.g. with `self.butler.registry.queryDatasets(...).count()`?